lua/lspconfig/configs.lua

```diff
@@ -255,35 +248,47 @@ function configs.__newindex(t, config_name, config_def)
           return
         end
       end
-      local buf_path = util.path.sanitize(bufname)
 
-      if get_root_dir then
-        root_dir = get_root_dir(buf_path, bufnr)
+      if project_root then
+        manager.add(project_root, false, bufnr)
+        return
       end
 
-      if root_dir then
-        manager.add(root_dir, false, bufnr)
-      elseif config.single_file_support then
-        local pseudo_root = #bufname == 0 and uv.cwd() or util.path.dirname(buf_path)
-        manager.add(pseudo_root, true, bufnr)
+      local buf_path = util.path.sanitize(bufname)
+
+      local function check_fast(root_dir, single_mode)
+        if not vim.in_fast_event() then
+          manager.add(root_dir, single_mode, bufnr)
+          return
+        end
+        vim.schedule(function()
+          manager.add(root_dir, single_mode, bufnr)
+        end)
       end
+
+      util.async_run(function()
+        local root_dir
+        if get_root_dir then
+          root_dir = get_root_dir(buf_path, bufnr)
```

Comment:
Because `get_root_dir` can be asynchronous, buffer could be deleted in the meantime
```suggestion
          root_dir = get_root_dir(buf_path, bufnr)
          if not api.nvim_buf_is_valid(bufnr) then
            return
          end
```